lib/core/src/Cardano/Wallet/Primitive/Migration/Selection.hs

```diff
@@ -0,0 +1,796 @@
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedLabels #-}
+{-# LANGUAGE OverloadedLists #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- |
+-- Copyright: © 2021 IOHK
+-- License: Apache-2.0
+--
+-- This module contains functions for incrementally constructing a selection
+-- to be included in a migration plan.
+--
+-- A selection is the basis for a single transaction.
+--
+-- Use 'create' to create a selection with one or more inputs.
+-- Use 'extend' to extend a selection with an additional input.
+-- Use 'verify' to verify the correctness of a selection.
+--
+module Cardano.Wallet.Primitive.Migration.Selection
+    (
+    -- * Types
+      Selection (..)
+    , SelectionCorrectness (..)
+    , SelectionError (..)
+    , SelectionFullError (..)
+    , RewardWithdrawal (..)
+
+    -- * Creating selections
+    , create
+
+    -- * Extending selections
+    , extend
+
+    -- * Verifying selections for correctness
+    , verify
+
+    -- * Computing bulk properties of selections
+    , computeCurrentFee
+    , computeCurrentSize
+    , computeMinimumFee
+
+    -- * Adding value to outputs
+    , addValueToOutputs
+
+    -- * Minimizing fees
+    , minimizeFee
+    , minimizeFeeStep
+
+    -- * Classes
+    , TxSize (..)
+
+    ) where
+
+import Prelude
+
+import Cardano.Wallet.Primitive.Types.Coin
+    ( Coin (..) )
+import Cardano.Wallet.Primitive.Types.TokenBundle
+    ( TokenBundle (..) )
+import Cardano.Wallet.Primitive.Types.TokenMap
+    ( TokenMap )
+import Cardano.Wallet.Primitive.Types.Tx
+    ( TxConstraints (..)
+    , txOutputCoinCost
+    , txOutputHasValidSize
+    , txOutputHasValidTokenQuantities
+    )
+import Control.Monad
+    ( (>=>) )
+import Data.Bifunctor
+    ( first )
+import Data.Either.Extra
+    ( eitherToMaybe, maybeToEither )
+import Data.Generics.Internal.VL.Lens
+    ( view )
+import Data.Generics.Labels
+    ()
+import Data.List.NonEmpty
+    ( NonEmpty (..) )
+import Data.Maybe
+    ( catMaybes, listToMaybe )
+import GHC.Generics
+    ( Generic )
+
+import qualified Cardano.Wallet.Primitive.Types.Coin as Coin
+import qualified Cardano.Wallet.Primitive.Types.TokenBundle as TokenBundle
+import qualified Cardano.Wallet.Primitive.Types.TokenMap as TokenMap
+import qualified Data.Foldable as F
+import qualified Data.List as L
+import qualified Data.List.NonEmpty as NE
+import qualified Data.Set as Set
+
+--------------------------------------------------------------------------------
+-- Selections
+--------------------------------------------------------------------------------
+
+data Selection i s = Selection
```

Comment:
Perhaps use more descriptive type parameters here.

Instead of `i`, use `inputId`.
Instead of `s`, use `size`.

Reply:
I think `size` makes sense if keeping it 👍 

How about just `input` instead of `inputId`? I don't get why this distinction is preferable.

Reply:
Okay, I've gone for `input` and `size`, as suggested.

Fixed in 2a2f17dc8384a68ff2e968f39a125d0a02f1e926.

> It seems i ~ (TxIn, TxOut) for the real algorithm also... Doesn't seem like an "id"?

Thinking about this, `createPlan` actually only needs to return a `MigrationPlan` where `i ~ TxIn`, since the caller of `createPlan` can figure out which `TxOut` each `TxIn` refers to, if they really need to do so.